### PR TITLE
Add eval plugin based on judge0 API

### PR DESCRIPTION
**Pull request by Ell**
This is kinda hacky and theres probably a better way to do polling, suggestions welcomed.

### CODE REVIEW

File: plugins/eval.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    try:
+        result = http.get_json(url, get_method="GET")
+    except http.HTTPError as e:
+        return None
+
+    status = result.get("status", {"id": 2})
+    status_id = status["id"]
+
+    # Processing, try again
+    if status_id not in [3, 6, 11]:
```
Comment (rmmh) [inline plugins/eval.py]: Let's map these status codes to descriptions using (lazily fetched) `https://api.judge0.com/statuses`. There are a lot of other error codes, this should probably be checking `if status_id in [1, 2]` instead (In Queue or Processing).

File: plugins/eval.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        return None
+
+    # Compilation error
+    if status_id == 6:
```
Comment (rmmh) [inline plugins/eval.py]: map error codes other than accepted using the descriptions from statuses

File: plugins/eval.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        return "[time: {time}, memory: {memory}] >> {stdout}".format(**result)
+
+
+def submit_code(language, code):
+    current_try = 0
+    output = None
+
+    url = "https://api.judge0.com/submissions?base64_encoded=true"
+
+    encoded_code = base64.b64encode(code.encode())
```
Comment (rmmh) [inline plugins/eval.py]: Why is this necessary? What JSON-invalid text are we receiving from IRC?

File: plugins/eval.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    except http.HTTPError as e:
+        return e
+
+    token = result.get("token")
+
+    if not token:
+        return "Missing submission token in response. API is probably down."
+
+    while current_try <= 5:
+        time.sleep(1)
```
Comment (rmmh) [inline plugins/eval.py]: Only waiting for 5 seconds seems a bit optimistic. Maybe use

```
for n in range(10):
    time.sleep(1.3 ** n)
```
which will wait for up to 42 seconds total.

File: plugins/eval.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+@hook.command("eval")
+def runcode(inp):
+    inputs = inp.split(" ")
+
+    arg1 = inputs[0]
+
+    if arg1 not in languages.keys():
+        return "Language not supported, supported languages: {}".format(
+            ", ".join(languages.keys())
```
Comment (rmmh) [inline plugins/eval.py]: `sorted(languages.keys())`

File: plugins/eval.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    if output:
+        return output
+
+    return "API took to long to return a response. Try again later."
+
+
+@hook.command("eval")
+def runcode(inp):
+    inputs = inp.split(" ")
```
Comment (rmmh) [inline plugins/eval.py]:
```
try:
    arg1, arg2 = inp.split(None, 1)
except ValueError:
    return usage()
```

File: plugins/eval.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+        current_try += 1
+
+    if output:
+        return output
+
+    return "API took to long to return a response. Try again later."
+
+
+@hook.command("eval")
```
Comment (rmmh) [inline plugins/eval.py]: add `autohelp=False` so we can provide a dynamic usage string

File: plugins/eval.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+        if output:
+            break
+
+        current_try += 1
+
+    if output:
+        return output
+
+    return "API took to long to return a response. Try again later."
```
Comment (rmmh) [inline plugins/eval.py]: "took too long"

File: plugins/eval.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
+import base64
+import time
+
+from util import hook, http
+
+
+languages = {
```
[the remark this answers is not in the document]
Comment (Ell) [reply plugins/eval.py]: Not sure this is going to work out, theres a lot of versions of languages that I assume we're just not gonna care about (like python 2.6 vs 2.7). Having to specify language + version seems like bad UX for how this plugin is going to get used.

File: plugins/eval.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        time.sleep(1.3 ** n)
+
+    if output:
+        return output
+
+    return "API took to long to return a response. Try again later."
+
+
+@hook.command("eval")
+def runcode(inp, autohelp=False):
```
Comment (rmmh) [inline plugins/eval.py]: autohelp is a parameter on hook.command

File: plugins/eval.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    m["clj"] = m["clojure"]
+    m["cpp"] = m["c++"]
+    m["csharp"] = m["c#"]
+    m["node"] = m["js"] = m["javascript"]
+    m["rb"] = m["ruby"]
+
+    return m
+
+
+languages = fetch_languages()
```
Comment (rmmh) [inline plugins/eval.py]: initialize languages and statuses inside of runcode, not at global scope. importing the module shouldn't cause network activity

File: plugins/eval.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+statuses = {x["id"]: x for x in http.get_json("https://api.judge0.com/statuses")}
+
+
+def get_result(token):
+    url = "https://api.judge0.com/submissions/{}".format(token)
+
+    try:
+        result = http.get_json(url, get_method="GET")
```
Comment (rmmh) [inline plugins/eval.py]: `get_method="GET"` ??

File: plugins/eval.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+        return "{} {}".format(status_description, stderr)
+
+    # Processing, try again
+    if status_id in [1, 2]:
+        return None
+
+    # Accepted and completed, return result and profiler stats
+    if status_id == 3:
+        return "[time: {time}, memory: {memory}] >> {stdout}".format(**result)
```
Comment (rmmh) [inline plugins/eval.py]: maybe "({time}s {memory}KB) {stdout}"

File: plugins/eval.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    supported_languages = ", ".join(sorted(languages.keys()))
+
+    try:
+        arg1, arg2 = inp.split(None, 1)
+    except ValueError:
+        return "eval <language> <code> - evaluate given code using language. Supported languages: {}".format(
+            supported_languages
+        )
+
+    if arg1 not in languages.keys():
```
Comment (rmmh) [inline plugins/eval.py]: `if arg1 not in languages:`

File: plugins/eval.py (hunk 8 above)
Comment (rmmh) [reply plugins/eval.py]: please fix

File: plugins/eval.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    return "API took to long to return a response. Try again later."
+
+
+@hook.command("eval", autohelp=False)
+def runcode(inp):
+    global languages
+    global statuses
+
+    if not languages:
+        languages = fetch_languages()
```
Comment (rmmh) [inline plugins/eval.py]: could do `languages.update(fetch_languages())` to avoid `global`

File: plugins/eval.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+    if not statuses:
+        statuses = {
+            x["id"]: x for x in http.get_json("https://api.judge0.com/statuses")
```
Comment (rmmh) [inline plugins/eval.py]: nothing is using statuses-- maybe omit this and change the above checks from using numeric status_id to using strings like `if status_description in ("In Queue", "Processing")`?

**Comment by rmmh** [comment]
Thanks!
